cpp/include/cudf/utilities/span.hpp

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cstddef>
+#include <limits>
+#include <type_traits>
+
+namespace cudf {
+namespace detail {
+
+constexpr size_t dynamic_extent = std::numeric_limits<std::size_t>::max();
+
+/**
+ * @brief C++20 std::span with reduced feature set.
+ */
+template <typename T, size_t Extent, typename Derived>
```

Comment:
It doesn't seem like CRTP is really necessary here. There isn't an interface that's being enforced.

Reply:
Using CRPT so we don't have to copy/paste any logic. Is there a better way to do that?

Reply:
You can just simply have `device_span` and `host_span` inherit from `span_base`. You use CRTP when you want to enforce an interface in the derived classes. Here you just want to avoid code duplication.

Reply:
We're returning the derived types in these functions. Is there a way to do this without CRPT and without defining those functions in the derived classes?

```c++
constexpr Derived first(size_type count) const noexcept;
constexpr Derived last(size_type count) const noexcept;
constexpr Derived subspan(size_type offset, size_type count) const noexcept;
```

Reply:
yes, it's called "covariant return type"!  (TIL). https://www.learncpp.com/cpp-tutorial/12-2a-the-override-and-final-specifiers-and-covariant-return-types/

Note I think it requires you to implement the function in the derived class -- you can't reuse the implementation from the base class.

Reply:
I wasn't aware c++ any covariant support! Interestingly. Interestingly, c++ doesn't support covariant type parameters (afaict), and so we can't use them with smart pointers. :(

Which is one of the use cases of CRTP.

https://www.fluentcpp.com/2017/09/12/how-to-return-a-smart-pointer-and-use-covariance/

But in this case, CRPT is allowing us to keep the shared logic in the base class.

Reply:
fwiw, this is the _very first time_ I've seen C#'s type engine provide a built-in feature that C++ template do not. It is required to suppose extension methods, as well as the TPL (task programming library), iirc.

https://docs.microsoft.com/en-us/dotnet/standard/generics/covariance-and-contravariance

@codereport you might find this thread interesting, if you don't already know all of it. :)

Reply:
I missed that some of the base class members needed to return a `Derived` type. In that case, CRTP makes sense here.